### PR TITLE
Logging using fluentd

### PR DESCRIPTION
**Pull request by lrog**
Implemented logging functionality for Docker-based deployments using Fluentd driver. When deployed, Fluend will be running as an additional microservice in order to collect and filter logs from the ones running in CogStack ecosystem.

Fluentd uses configuration files to define the filtering and output rules for messages coming from predefined source(s). In our current configuration, the Fluentd driver will be listening at `localhost` on port `24224` for the incoming messages, as defined both in the Docker Compose file and Fluentd configuration files. The default configuration file for Fluentd is `docker-cogstack/fluentd/conf/fluent.conf`.

In our setup, Fluentd uses additional filter plugins hence a custom Fluentd image is used. This image will be available to download directly from CogStack Dockerhub under the name `cogstacksystems/fluentd` (or while still in development in `dev` branch: `cogstacksystems/fluentd:dev-latest` ). Alternatively, the image can be build locally using the Dockerfile in the directory `docker-cogstack/fluentd` in the CogStack repository.

A full working example i.e., Example 6 has been created (based on Example 2) to demonstrate the logging functionality -- see: `examples/example6`. The general rule applied is that all the messages which have been previously sent to `stderr` will be classified as error messages, where as the ones sent to `stdout` -- as informative ones.  The logging mechanisms have been covered in the examples documentation in `docs/examples`.

After filtering and parsing, all the messages are output into files, separate per each service tag. The logs containing only error messages are stored in files starting with `erorr.*` prefix, whereas the full logs are stored in files starting with `full.*` prefix. The log files are managed using log rotating policy, keeping new logs for 24 hours, before being archived. 

As a side note, since the error messages are of high importance, on arrival they are additionally printed to the standard output for the user's instant inspection.

The logs are output to files in JSON format and they contain such fields:
- `container_id` and `container_name` of the running microservice,
- `log` message,
- `time` of the message arrival.

### CODE REVIEW

**Comment by yatharthranjan** [comment]
This is great work. Just a minor general observation that we should consider is that -- if we enable third party plugins for logging, the `docker logs` command as well as the docker logs in Docker API will start failing and there will be no output. see - https://github.com/moby/moby/issues/30887
In RADAR too we were thinking of adding this but did not because of the above reason, and since we also have portainer (which depends on Docker API) so that would also stop displaying logs. We will always need to parse the files in-order to see the logs which may not be more convenient than the quick `docker logs`

**Comment by afolarin** [comment]
:+1:  @yatharthranjan -- insightful as ever. Yes, looks like bifurcating the logs it's available but not supported on the community edition of Docker.. :disappointed: 
I think for this usecase it's probably more informative to most to have the logs out to fluentd, but I'd like to have this configurable in the stack.

File: docs/examples/index.md [hunk 1]
```diff
@@ -1237,3 +1257,69 @@ The *properties* file used in this example is similar to the one from [Example 1
...
+# <a name="example-6"></a> Example 6
+[//]: # "-------------------------------------------------------------------------------------"
+
+## General information
+
+This example is an extension of [Example 2](#example-2) providing logging mechanism using [Fluentd](https://www.fluentd.org/) log collector and it only focuses on the logging part.
+
+
+[//]: # "<span style='color:red'> NOTE: </span>"
+**Note: For the moment, the Docker containers used in this example: `cogstack-pipeline` and `fluentd` are not yet available to pull the official CogStack Dockerhub repository. These images need to be build locally. The scripts to build these images are available in the main CogStack bundle in the directory `localBuild/`.**
```
Comment (afolarin) [inline docs/examples/index.md]: update, they are now

**Review by afolarin — APPROVED** [review]
LGTM

only some minor tweaks to the container builds which are now changed

**Comment by lrog** [comment]
Thanks @yatharthranjan for the hint! 👍 Let's then keep an eye on the future Docker CE edition features and whether we could use both docker logs functionality alongside fluentd or any other logging drivers. fluentd can be quite useful service in our configurable stack and when we see that it can be a bit limiting then we could either modify, exchange it or use directly docker logs.

As for parsing the log files, you are right -- in the current implementation fluentd logs are output in JSON format, however `jq` becomes here pretty handy 😉 I hope that quite useful should be now filtering the high priority messages displayed to the console.

File: docs/examples/index.md (hunk 1 above)
Comment (lrog) [reply docs/examples/index.md]: 👍 Update in aea0deb

**Pull request merged**